crates/valence/src/inventory/validate.rs

```diff
@@ -0,0 +1,675 @@
+use anyhow::ensure;
+use valence_protocol::packet::c2s::play::click_slot::ClickMode;
+use valence_protocol::packet::c2s::play::ClickSlotC2s;
+
+use super::{Inventory, InventoryWindow, PLAYER_INVENTORY_MAIN_SLOTS_COUNT};
+use crate::prelude::CursorItem;
+
+/// Validates a click slot packet enforcing that all fields are valid.
+pub(crate) fn validate_click_slot_impossible(
+    packet: &ClickSlotC2s,
+    player_inventory: &Inventory,
+    open_inventory: Option<&Inventory>,
+) -> anyhow::Result<()> {
+    ensure!(
+        (packet.window_id == 0) == open_inventory.is_none(),
+        "window id and open inventory mismatch: window_id: {} open_inventory: {}",
+        packet.window_id,
+        open_inventory.is_some()
+    );
+
+    let max_slot = match open_inventory {
+        Some(inv) => inv.slot_count() + PLAYER_INVENTORY_MAIN_SLOTS_COUNT,
+        None => player_inventory.slot_count(),
+    };
+
+    // check all slot ids and item counts are valid
+    ensure!(
+        packet.slots.iter().all(|s| {
+            (0..=max_slot).contains(&(s.idx as u16))
+                && s.item
+                    .as_ref()
+                    .map_or(true, |i| (1..=127).contains(&i.count()))
```

Comment:
The `1..=127` range fixes the issue with the weird stack counts, but now clients can stack items in ways they shouldn't.

You can check the max stack count with `ItemKind::max_stack`. For handling the weird counts, I think you need to check that the new stack does not have a count greater than `max(old_stack.item.max_stack(), old_stack.count())` or something like that.